Do not reference PTRACE_{G,S}ETFPXREGS when undefined

umarcor opened the pull request:
Issue: #3381
Issue: #3399

Signed-off-by: Unai Martinez-Corral <unai.martinezcorral@ehu.eus>

---

Building on `armv7l` or `armv8l` fails with:

``` bash
[ 64%] Building C object core/CMakeFiles/drinjectlib.dir/config.c.o
/dynamorio/core/unix/injector.c:866:50: error: 'PTRACE_GETFPXREGS' undeclared here (not in a function)
                                                { PTRACE_GETFPXREGS, "PTRACE_GETFPXREGS" },
                                                  ^~~~~~~~~~~~~~~~~
/dynamorio/core/unix/injector.c:867:50: error: 'PTRACE_SETFPXREGS' undeclared here (not in a function)
                                                { PTRACE_SETFPXREGS, "PTRACE_SETFPXREGS" },
                                                  ^~~~~~~~~~~~~~~~~
make[2]: *** [core/CMakeFiles/drinjectlib.dir/unix/injector.c.o] Error 1
```

This PR avoids using `PTRACE_GETFPXREGS` and `PTRACE_SETFPXREGS` on `ARM`, as it is currently done on `AARCH64`.

derekbruening commented:
Certainly our ARM build on Travis works fine with the ptrace defines as they are, so this seems more like a toolchain version issue.  Please specify which toolchain needs this change.

umarcor commented:
I am not sure about how to do that. I found the problem on Ubuntu 18.04 with `apt install -y cmake g++-6 gcc`.

egrimley commented:
I get the impression that `PTRACE_GETFPXREGS` and `PTRACE_SETFPXREGS` are perhaps obsolete and not available on newer systems, though I can't find an explanation of why. Online man pages for ptrace don't mention those symbols.

If you want to build with GCC 7, replace `||` with `|` in line 1888 of `opnd_shared.c`. (That bug should be fixed in any case. I thought it had been.)

egrimley commented:
This patch looks good to me, though the commit message should be improved.

There's the obvious alternative of replacing the architecture test with a direct test for the macro in question:
```
#    ifdef PTRACE_GETFPXREGS
                                               { PTRACE_GETFPXREGS, "PTRACE_GETFPXREGS" },
#    endif
```

I tried grepping the Linux kernel source and found `FPXREGS` mentioned only under `arch/mips/` and `arch/x86/`, so perhaps the test should be `#ifdef X86`, if it is going to be an architecture test.

umarcor commented:
@egrimley I was trying to find some "solid" proof that `PTRACE_GETFPXREGS` and `
PTRACE_SETFPXREGS` are specific to Intel. In the sources of the latest kernel they are only found under `arch/x86` (https://elixir.bootlin.com/linux/v4.20.13/ident/PTRACE_GETFPXREGS), but I don't know if that is enough to defend merging the PR as is.

umarcor commented:
> This patch looks good to me, though the commit message should be improved.

I changed it.

> There's the obvious alternative of replacing the architecture test with a direct test for the macro in question:

I implemented this now. It both macros exist, both are added. Otherwise, none.

umarcor commented:
@derekbruening, which should the commit/title if there is no specific issue number to fix by this PR?

derekbruening commented:
> @derekbruening, which should the commit/title if there is no specific issue number to fix by this PR?

Just a short description.  "Do not reference PTRACE_{G,S}ETFPXREGS when undefined" or some such.

> umarcor force-pushed the umarcor:fix-ptrace-fpxregs  branch 3 times

Please don't force-push: it deletes past review comments and makes it impossible for the reviewer to see what changed since the last review.  The squash-and-merge cleans up the final commit.  (See https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member, https://github.com/DynamoRIO/dynamorio/wiki/Contributing)

umarcor commented:
> Please don't force-push: it deletes past review comments and makes it impossible for the reviewer to see what changed since the last review. The squash-and-merge cleans up the final commit. (See [DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member](https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member), [DynamoRIO/dynamorio/wiki/Contributing](https://github.com/DynamoRIO/dynamorio/wiki/Contributing))

Ok. Sorry for the inconvenience. Just to make it clear:

- You don't mind the actual commit messages, because you are going to squash and use the title of the PR as the final message. Therefore:
  - I should not edit already pushed commits to fix the message.
  - I should neither rebase my branches on top of master.

Is it correct?

derekbruening commented:
> > Please don't force-push: it deletes past review comments and makes it impossible for the reviewer to see what changed since the last review. The squash-and-merge cleans up the final commit. (See [DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member](https://github.com/DynamoRIO/dynamorio/wiki/Code-Reviews#submitting-a-change-as-a-non-project-member), [DynamoRIO/dynamorio/wiki/Contributing](https://github.com/DynamoRIO/dynamorio/wiki/Contributing))
> 
> Ok. Sorry for the inconvenience. Just to make it clear:
> 
> - You don't mind the actual commit messages, because you are going to squash and use the title of the PR as the final message. Therefore:

The PR title would be the default commit title (== first line), yes, but the default final commit body (== lines beyond the first line) comes from the commit message bodies concatenated together.  Whoever does the squash merges those messages together in some fashion, usually by discarding later bug fix or tweak commit messages.

>   - I should not edit already pushed commits to fix the message.
>   - I should neither rebase my branches on top of master.

Right, in general do not change the history of a shared git branch: rebase/history rewrite is for private branches.

I wish Github would have first-class support for the final commit message and title being some separate field that would take review comments and be editable, but given that support not being there, one method is to make the next commit's message be the new full message for the final squash.  The squasher would then take that one and discard the earlier one.

derekbruening commented on core/unix/injector.c:
```diff
@@ -862,7 +862,7 @@ static const enum_name_pair_t pt_req_map[] = { { PTRACE_TRACEME, "PTRACE_TRACEME
                                                { PTRACE_ATTACH, "PTRACE_ATTACH" },
                                                { PTRACE_DETACH, "PTRACE_DETACH" },
-#    ifndef AARCH64
+#    if defined(PTRACE_GETFPXREGS) && defined(PTRACE_SETFPXREGS)
```
(I thought I sent this out a long time ago but it was sitting pending)

umarcor replied:
> nit: It seems cleaner to have a separate define check for each of these, rather than bundling them, right?

I thought about it. But, it would be really strange to me if one of them was defined but not the other. I assume that both need to come together, so bundling them is meant to reduce verbosity.

Nevertheless, if you find it more clear to have them separated, I can update it ASAP.

derekbruening replied:
OK either way.  Will just merge it as-is.

derekbruening commented:
run arm tests

derekbruening commented:
Looks like #2898 but sigsuspend has never been flaky on x86 before.
